018-cli-2.0/018-cli-2.0.md

```diff
@@ -0,0 +1,312 @@
+- Dependencies on libraries, tools, projects or work that isn't yet complete
+- New API routes that need to be created or modifications to the existing routes (if needed)
+- Any breaking changes and ways in which we can ensure backward compatibility.
+- Use Cases
+- Goals
+- Deliverables
+- Changes to documentation
+- Ways to scale the solution
+
+Ensure that you include examples, code-snippets etc. to allow the community to understand the proposed solution. **It would be best if the examples use naming conventions that you intend to use during the actual implementation so that changes can be suggested early on during the development.**
+
+Write your answer below.
+
+-->
+
+This refactor of the CLI will revolve around the following areas
+1. Migration of the CLI from a Server Side SDK to a Console SDK
+2. Easier deployment of Cloud Functions
+3. Easier creation of Cloud Functions
+
+
+## Migration of the CLI from a Server SDK to a Console SDK
+---
+ 
+Until now, the CLI was based off of our Server Side Swagger Spec. While this was a good start, it limits our functionality to a Server SDK. Ideally we want the CLI to be an alternative to the Console SDK ( aka the Appwrite Dashboard ). Here are some changes that we will need to make to adhere to the server spec.
+
+After the implementation of this RFC, the CLI will behave more like the Appwrite Console rather than a Server SDK. This accounts for a new form of Authentication that is similar to the way we handle logins in the Appwrite console. We use a cookie.
+
+We will introduce a new command `appwrite login` that will work in one of either ways.
+
+* `appwrite login` makes a request to the existing `accounts.createSession` endpoint, captures the returned cookie and uses this cookie in all the subsequent requests. 
+
+    Whenever we wish to access console specific features, we can authenticate those requests using the cookie and continue to use the API Key based authentication when using the CI.
+
+    ```sh
+    $ appwrite login
+    Email:
+    Password:
+    ```
+
+    In CI environments, the authentication will continue to take place using the API key on a per project basis until we develop a new token / key based mechanism for console authentication. 
+
+    ```sh
+    appwrite client --setKey="" --setEndpoint="" ...
+    ```
+
+* Second method relies on a browser based auth.
+
+  ```sh
+  $ appwrite login
+
+  # Browser based authentication continues. After a successful login, the returned token and user ID are stored in a hidden preferences file.
+  ```
+
+  1. The command first finds an available local port. 
+  2. Construct a redirect URL using the available port `https://localhost:1234`
+  3. Spin up a local http server listening to on the available port ( 1234 in this case )
+  4. Opens the browser with a request to either the local appwrite server or appwrite cloud `/login?callback=https://localhost:1234` with the redirect URL as one of the query parameters.
```

Comment:
```suggestion
  4. Opens the browser with a request to either the local appwrite server or appwrite cloud `/authorize/cli?callback=https://localhost:1234` with the redirect URL as one of the query parameters.
```

This will also allow us to use `/authorize/vscode?callback=https://localhost:1234` in the future. We should also list some of the changes that will be required in the console itself. For example:
1. Custom redirect after login - if the user is not logged in the console, he should be redirect to the signin page, and after the signin he should be brought back to `/authorize/cli?callback=https://localhost:1234`.

2. We should also add protection against [open-redirect attack](https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html).

Reply:
done :+1: